File: Graph/python/hasPath.py
from .graph_utils import Graph


def has_path(graph: Graph, src: str, dst: str) -> bool:
    if src == dst:
        return True
    visited: set[str] = set()
    return _dfs(graph, src, dst, visited)


def _dfs(graph: Graph, node: str, dst: str, visited: set[str]) -> bool:
    if node == dst:
        return True
    if node in visited:
        return False

    visited.add(node)
    for neighbor in graph.get(node, []):
        if _dfs(graph, neighbor, dst, visited):
            return True
    return False

File: Graph/ts/hasPath.ts
import { Graph } from "./types";

export function hasPath(graph: Graph, src: string, dst: string): boolean {
  if (src === dst) return true;
  const visited = new Set<string>();
  return dfs(graph, src, dst, visited);
}

function dfs(graph: Graph, node: string, dst: string, visited: Set<string>): boolean {
  if (node === dst) return true;
  if (visited.has(node)) return false;
  visited.add(node);

  for (const neighbor of graph[node] ?? []) {
    if (dfs(graph, neighbor, dst, visited)) {
      return true;
    }
  }
  return false;
}
